### ISSUE TITLE
Why concatenate after norm ? 

### ISSUE THREAD

**Issue by Medabid1** [chrischute/real-nvp#3]
Hello, 
could you plz explain, why double the input after the batchNorm and concat [input, -input]? 

thanks

**Comment by chrischute** [comment]
This only happens for checkerboard coupling, because half of the inputs get zero-ed out by the checkerboard mask. Doubling re-scales the variance so the batchnorm operation is the same as it would be if you only considered the non-zeroed-out inputs.

The corresponding code released by the authors of the paper can be found [here](https://github.com/tensorflow/models/blob/master/research/real_nvp/real_nvp_multiscale_dataset.py#L321-L327).

**Comment by KeAWang** [comment]
@chrischute Do you remember why you concatenate `x` and `-x`? Also do you know why the authors also concatenate `x` (`res` in their code) and the mask?

**Comment by chrischute** [comment]
I was following the [original paper's code](https://github.com/tensorflow/models/blob/master/research/real_nvp/real_nvp_multiscale_dataset.py#L321-L327) as closely as possible, just porting from TensorFlow to PyTorch. I'm not sure why they concatenate x and -x... It might be a trick to make the mean 0. I never understood why they concatenate res and mask. If you ask the original authors I bet they'd have an answer.

**Comment by KeAWang** [comment]
Got it. Thanks!
